Add custom caveat enforcer tutorial (#2253)
MetaMask/metamask-docs: delegation-toolkit/tutorials/create-custom-caveat-enforcer.md

---

> description: Follow this tutorial to create a custom caveat enforcer for a delegation.

Reviewer: Just a question why we are changing the sentence formation? Also, we should target deploy, and apply keywords as well.

# Create a custom caveat enforcer

This tutorial walks you through creating a custom [caveat enforcer](../concepts/delegation/caveat-enforcers.md) and applying it to a [delegation](../concepts/delegation/index.md).

The MetaMask Delegation Toolkit includes [out-of-the-box caveat enforcers](../reference/caveats.md) that define rules and restrictions for common use cases.
For more specific control or other use cases, you can create custom caveat enforcers.
In this tutorial, you'll create and apply a caveat enforcer that only allows a delegation to be redeemed after a specific timestamp.

## Prerequisites

- [Install and set up the Delegation Toolkit](../get-started/install.md) in your project.
- [Configure the Delegation Toolkit.](../guides/configure.md)
- [Install Foundry and Forge.](https://getfoundry.sh/introduction/installation)
- Get an [Infura API key](/developer-tools/dashboard/get-started/create-api) from the MetaMask Developer dashboard.
- Have a MetaMask account with some Sepolia ETH to deploy your contract.
  :::note
  You can use the [MetaMask faucet](/developer-tools/faucet) to get Sepolia ETH.
  :::

## Steps

### 1. Create the caveat enforcer

> In your project's `src` directory, create a contract that extends the

Reviewer: Let's say create a new directory `contracts`, and there create a new contract.

[`ICaveatEnforcer.sol`](https://github.com/MetaMask/delegation-framework/blob/main/src/interfaces/ICaveatEnforcer.sol)
interface.

> The following `AfterTimestampEnforcer.sol` caveat enforcer only allows a delegation to be redeemed after a specific timestamp:

Reviewer: Or we say that create a new contract with AfterTimestampEnforcer since it's a tutorial.

> The following `AfterTimestampEnforcer.sol` caveat enforcer only allows a delegation to be redeemed after a specific timestamp:

Reviewer: We also need to explain the rational behind only implementing the beforeHook, and why it's important. We can create an issue for improvement, I can add while doing changes for v0.13.

> The following `AfterTimestampEnforcer.sol` caveat enforcer only allows a delegation to be redeemed after a specific timestamp:

Author: raised issue https://github.com/MetaMask/metamask-docs/issues/2257

# Create a custom caveat enforcer

This tutorial walks you through creating a custom [caveat enforcer](../concepts/caveat-enforcers.md) and applying it to a [delegation](../concepts/delegation.md).

> The MetaMask Delegation Toolkit provides some [out-of-the-box caveat enforcers](../reference/caveats.md) that define rules and restrictions for common use cases.

Reviewer 2: Suggestion, something like, "The MetaMask Delegation Toolkit includes [built-in caveat enforcers](https://github.com/MetaMask/metamask-docs/pull/reference/caveats.md) for common use cases. For more specific control, you can create custom caveat enforcers."?

Author: I made minor edits here, but not sure if "built-in" is the right term to use for the caveat enforcers, as it might imply that they are applied by default?

For more granular or custom control, you can create custom caveat enforcers from scratch.

> This tutorial walks you through creating and applying a caveat enforcer that only allows a delegation to be redeemed after a specific timestamp.

Reviewer 2: We mention, "This tutorial walks you..." twice. Should probably be combined.

### 2. Deploy the caveat enforcer

Deploy your custom caveat enforcer using [Forge](https://book.getfoundry.sh/forge/deploying) to obtain its contract address.
Replace `<YOUR-API-KEY>` with your Infura API key, and `<YOUR-PRIVATE-KEY>` with the private key of your MetaMask account:

```bash
forge create src/AfterTimestampEnforcer.sol:AfterTimestampEnforcer \
  --rpc-url https://sepolia.infura.io/v3/<YOUR-API-KEY>            \
  --private-key <YOUR-PRIVATE-KEY>                                 \
  --broadcast
```

The Forge CLI will display the address of the deployed caveat enforcer.

### 3. Apply the caveat enforcer

> Specify the address where your `AfterTimestampEnforcer.sol` contract is deployed, add it to the caveat builder, and create a delegation.

Reviewer 2: "Specify the address of the deployed `AfterTimestampEnforcer.sol` contract"

Learn more about [applying caveats to a delegation](../guides/create-delegation/restrict-delegation.md).

> The following code snippet uses the custom caveat enforcer to create a delegation granting
> an allowance of 1,000,000 wei that can only be spent after one hour from when the delegation is created:

Reviewer 2: ```suggestion
The following code snippet uses the custom caveat enforcer to create a delegation with a 1,000,000 wei allowance that becomes spendable 1 hour after it is created:
```
